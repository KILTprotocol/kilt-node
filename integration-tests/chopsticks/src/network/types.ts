import type { setupContext } from '@acala-network/chopsticks-testing'

export type Config = Awaited<ReturnType<typeof setupContext>>
