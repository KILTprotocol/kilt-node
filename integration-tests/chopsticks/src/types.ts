export interface AssetSwitchSupplyParameters {
	circulatingSupply: bigint
	sovereignSupply: bigint
	totalSupply: bigint
}
