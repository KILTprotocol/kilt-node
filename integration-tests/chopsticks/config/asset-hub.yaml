endpoint:
  - wss://asset-hub-rococo-rpc.dwellir.com
mock-signature-host: true
runtime-log-level: 5
port: 50001
db: db/asset-hub.sqlite

import-storage:
  System:
    Account:
      -
        -
          - 5Go9qhpqq2EzAZ1SKkkx6bTjnSoTnHEB8FHCnieNS57TimEF
        - providers: 1
          data:
            free: '100000000000000000000000'
