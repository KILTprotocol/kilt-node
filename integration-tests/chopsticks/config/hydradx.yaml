endpoint:
 - wss://hydradx-rpc.dwellir.com
 - wss://rpc.hydradx.cloud
db: ./db/hydradx.db.sqlite
port: 50002
runtime-log-level: 5

import-storage:
 TechnicalCommittee:
  Members:
   - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
 Council:
  Members:
   - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
 System:
  Account:
   -
    -
     - 4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr
    - providers: 1
      data:
       free: "1000000000000000000"
   -
    -
     - 4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ
    - providers: 1
      data:
       free: "1000000000000000000"
 AssetRegistry:
  Assets:
   -
    -
     - 28
    - name: KILT
      assetType: Token
      existentialDeposit: 21358393848783
      symbol: KILT
      decimals: 15
      location:
       parents: 1
       interior: X1(Parachain(2086))
      isSufficient: true
  AssetIds:
   -
    -
     - KILT
    - 28
  AssetLocations:
  -
   -
    - 28
   - parents: 1
     interior:
      X1:
       Parachain: 2086
  LocationAssets:
   -
    -
     - parents: 1
       interior:
        X1:
         Parachain: 2086
    - 28
 MultiTransactionPayment:
  AcceptedCurrencies:
   -
    -
     - 28
    - "72639897479709500000"

