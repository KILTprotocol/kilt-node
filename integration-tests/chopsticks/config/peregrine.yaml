endpoint:
  - wss://peregrine.kilt.io
mock-signature-host: true
db: db/peregrine.sqlite
runtime-log-level: 5
wasm-override: /home/antonio/Developer/kilt-node/target/debug/wbuild/peregrine-runtime/peregrine_runtime.compact.compressed.wasm
