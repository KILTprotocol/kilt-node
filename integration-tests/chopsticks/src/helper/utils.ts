import { ApiPromise, SubmittableResult } from '@polkadot/api'
import { Keyring } from '@polkadot/keyring'
import { u8aToHex } from '@polkadot/util'
import { decodeAddress } from '@polkadot/util-crypto'

import type { SubmittableExtrinsic } from '@polkadot/api/types'
import type { KeyringPair } from '@polkadot/keyring/types'
import type { Codec } from '@polkadot/types/types'
import type { ExpectStatic } from 'vitest'
import { EventRecord } from '@polkadot/types/interfaces'

/**
 * Keyring instance to generate key pairs with 38 ss58Format encoding.
 */
const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

/**
 * Key pairs for Alice, Bob, and Charlie used for testing.
 */
export const keysAlice = keyring.addFromUri('//alice', undefined)
export const keysBob = keyring.addFromUri('//bob', undefined)
export const keysCharlie = keyring.addFromUri('//charlie', undefined)

/**
 * Converts a given value to a Number
 */
export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}
	return Number(value)
}

/**
 * Converts a given address to its hexadecimal representation.
 *
 * @param addr - The address to be converted. It should be a string representation of an address.
 */
export function hexAddress(addr: string) {
	return u8aToHex(decodeAddress(addr))
}

/**
 * Validates if the received balance is within the expected range considering a certain precision.
 *
 * @param previousBalance - The balance before the operation.
 * @param receivedBalance - The balance after the operation.
 * @param deltaBalance - The expected change in balance.
 * @param expect - The assertion function from the testing library.
 * @param precision - The precision of the balance validation. It must be a value between 0 and 100.
 *
 * The function calculates the expected balance by adding the deltaBalance to the previousBalance.
 * It then calculates the lower and upper bounds of the expected balance considering the precision.
 * The received balance is then checked if it falls within the range of the lower and upper bounds.
 *
 * If the precision is not between 0 and 100, the function throws an error.
 *
 * @throws {Error} If the precision is not between 0 and 100.
 */
export function validateBalanceWithPrecision(
	previousBalance: bigint,
	receivedBalance: bigint,
	deltaBalance: bigint,
	expect: ExpectStatic,
	precision: bigint
) {
	if (precision < 0n || precision > 100n) {
		throw new Error('Precision must be between 0 and 100')
	}

	const allowedError = 100n - precision
	const expectedBalance = previousBalance + deltaBalance

	const lowerBound = expectedBalance - (expectedBalance * allowedError) / 100n
	const upperBound = expectedBalance + (expectedBalance * allowedError) / 100n

	expect(receivedBalance).toBeGreaterThanOrEqual(lowerBound)
	expect(receivedBalance).toBeLessThanOrEqual(upperBound)
}

/**
 * Fetches the paid fees for the executed XCM message. Is only working on the sender chain.
 */
export async function getPaidXcmFees(api: ApiPromise, eventRecords: Codec[]): Promise<bigint> {
	const polkadotFees = eventRecords.find((eventRecord) =>
		api.events.polkadotXcm.FeesPaid.is((eventRecord as EventRecord).event)
	)
	if (!polkadotFees) {
		throw new Error('FeesPaid event not found')
	}

	// eslint-disable-next-line @typescript-eslint/no-explicit-any
	return (polkadotFees as any).event.data.fees[0].fun.asFungible.toBigInt()
}

/**
 * Calculates the transaction fees for a given transaction.
 */
export async function calculateTxFees(
	tx: SubmittableExtrinsic<'promise', SubmittableResult>,
	account: KeyringPair
): Promise<bigint> {
	const paymentInfo = await tx.paymentInfo(account)
	return paymentInfo.partialFee.toBigInt()
}

// Constants for the balances
export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)

export const initialBalanceKILT = 100n * KILT
export const initialBalanceDOT = 100n * DOT
export const initialBalanceHDX = 100n * HDX
