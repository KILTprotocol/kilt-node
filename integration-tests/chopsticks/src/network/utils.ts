export function getSiblingLocation(paraId: number) {
	return {
		parents: 1,
		interior: {
			X1: { Parachain: paraId },
		},
	}
}

export function getParentLocation() {
	return {
		parents: 1,
		interior: 'Here',
	}
}

export function getAccountLocationV2(addr: string) {
	return {
		V2: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						network: 'Any',
						id: addr,
					},
				},
			},
		},
	}
}

export function getAccountLocationV3(addr: string) {
	return {
		V3: {
			parents: 0,
			interior: {
				X1: {
					AccountId32: {
						id: addr,
					},
				},
			},
		},
	}
}

export function getNativeAssetIdLocation(amount: bigint) {
	return {
		id: { Concrete: { parents: 0, interior: 'Here' } },
		fun: { Fungible: amount },
	}
}
