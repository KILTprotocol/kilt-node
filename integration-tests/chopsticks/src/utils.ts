import { Keyring } from '@polkadot/keyring'

import { AssetSwitchSupplyParameters } from './types.js'

const keyring = new Keyring({ type: 'ed25519', ss58Format: 38 })

export const keysAlice = keyring.addFromUri('//alice', undefined, 'ed25519')
export const keysBob = keyring.addFromUri('//bob', undefined, 'ed25519')
export const keysCharlie = keyring.addFromUri('//charlie', undefined, 'ed25519')

export function toNumber(value: string | undefined): number | undefined {
	if (value === undefined) {
		return undefined
	}

	return Number(value)
}

export function getAssetSwitchParameters(
	totalSupply: bigint = BigInt('100000000000000000000'),
	ratioCirculating: bigint = BigInt(10)
): AssetSwitchSupplyParameters {
	const circulatingSupply = (totalSupply * ratioCirculating) / BigInt(100)
	const sovereignSupply = totalSupply - circulatingSupply
	return { circulatingSupply, sovereignSupply, totalSupply }
}

export const KILT = BigInt(1e15)
export const DOT = BigInt(1e10)
export const HDX = BigInt(1e12)
export const ROC = BigInt(1e12)

export const initialBalanceKILT = BigInt(100) * KILT
export const initialBalanceDOT = BigInt(100) * DOT
export const initialBalanceHDX = BigInt(100) * HDX
export const initialBalanceROC = BigInt(100) * ROC
