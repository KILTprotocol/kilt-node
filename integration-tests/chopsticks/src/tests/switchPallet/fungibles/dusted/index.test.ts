import { describe, beforeEach, it, afterEach } from 'vitest'
import { sendTransaction, withExpect } from '@acala-network/chopsticks-testing'
import type { KeyringPair } from '@polkadot/keyring/types'

import { createBlock, setStorage } from '../../../../network/utils.js'
import { calculateTxFees, hexAddress } from '../../../../helper/utils.js'
import { testCases } from './config.js'
import { Config } from '../../../../network/types.js'
import { setupNetwork, shutDownNetwork } from '../../../../network/utils.js'

describe.skip.each(testCases)('Dust account', { timeout: 30_000 }, async ({ account, query, txContext, config }) => {
	let context: Config
	let senderAccount: KeyringPair
	let receiverAccount: KeyringPair

	const { desc, network, storage } = config

	// Create the network context
	beforeEach(async () => {
		const { parachains, relay } = network

		const { parachainContexts } = await setupNetwork(relay, parachains)
		const [senderChainContext] = parachainContexts
		context = senderChainContext

		const { senderStorage } = storage
		await setStorage(context, senderStorage)

		senderAccount = account.sender
		receiverAccount = account.receiver
	}, 20_000)

	// Shut down the network
	afterEach(async () => {
		try {
			await shutDownNetwork([context])
		} catch (error) {
			if (!(error instanceof TypeError)) {
				console.error(error)
			}
		}
	})

	it(desc, { timeout: 10_000, retry: 3 }, async ({ expect }) => {
		const { checkSystemEvents } = withExpect(expect)
		const { balanceToTransfer, events, tx } = txContext

		// check initial state

		const nativeBalance = await query.native(context, senderAccount.address)
		const foreignBalance = await query.foreign(context, senderAccount.address)
		const receiverNativeBalance = await query.native(context, receiverAccount.address)

		expect(nativeBalance).toBe(balanceToTransfer)
		expect(foreignBalance).toBeGreaterThan(BigInt(0))
		expect(receiverNativeBalance).toBe(BigInt(0))

		let rawTx = tx(context, hexAddress(receiverAccount.address), balanceToTransfer.toString())
		const txFees = await calculateTxFees(rawTx, senderAccount)
		rawTx = tx(context, hexAddress(receiverAccount.address), (balanceToTransfer - txFees).toString())

		await sendTransaction(rawTx.signAsync(senderAccount))

		// process tx
		await createBlock(context)

		// check balance movement
		const nativeBalanceAfter = await query.native(context, senderAccount.address)
		const foreignBalanceAfter = await query.foreign(context, senderAccount.address)
		const receiverNativeBalanceAfter = await query.native(context, receiverAccount.address)

		expect(nativeBalanceAfter).toBe(BigInt(0))
		// sender should keep the foreign balance
		expect(foreignBalanceAfter).toBeGreaterThan(BigInt(0))
		expect(receiverNativeBalanceAfter).toBe(balanceToTransfer - txFees)

		events.map(
			async (pallet) =>
				await checkSystemEvents(context, pallet).toMatchSnapshot(
					`${desc}: Dusted accounts pallet: ${JSON.stringify(pallet)}`
				)
		)
	})
})
