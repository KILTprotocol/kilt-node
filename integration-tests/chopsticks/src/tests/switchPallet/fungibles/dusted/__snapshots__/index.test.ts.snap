// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Dust account > V4 LIVE > V4 LIVE: Dusted accounts pallet: {"section":"balances","method":"Endowed"} 1`] = `
[
  {
    "data": {
      "account": "4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr",
      "freeBalance": "(rounded 100000000000000000)",
    },
    "method": "Endowed",
    "section": "balances",
  },
]
`;

exports[`Dust account > V4 LIVE > V4 LIVE: Dusted accounts pallet: {"section":"balances","method":"Transfer"} 1`] = `
[
  {
    "data": {
      "amount": "(rounded 100000000000000000)",
      "from": "4seWojfEHrk5YKPahdErazQ3CWEHZYi6NV4gKz5AaejWbRPJ",
      "to": "4rsmbFBYpVmWE2LHRsSZKxf22a8cbJLxMZqvVGkGkSDmDBcr",
    },
    "method": "Transfer",
    "section": "balances",
  },
]
`;
